docs/contribute/testing.adoc

```diff
@@ -0,0 +1,423 @@
+rule from a specific profile instead of all the rules in that profile.
+
+NOTE: The single rule evaluation feature was added into the OpenSCAP through
+the commit
+link:https://github.com/OpenSCAP/openscap/commit/4637b132e710d7ab73601718a531a8587b406cd3[4637b13].
+
+To sum up, we are adding the highlighted option to the `xccdf` module of
+the OpenSCAP command line tool:
+[[app-listing]]
+[subs=+quotes]
+----
+$ oscap xccdf eval *--rule X* --profile Y \
+    /usr/share/xml/scap/ssg/content/ssg-fedora-ds.xml
+----
+
+
+=== Deciding where to put a new test
+In our example, we are changing the `xccdf` module, therefore we will look for
+its subdirectory in the link:../../tests[tests] direcotory and we will find it
+at the following link: link:../../tests/API/XCCDF[tests/API/XCCDF]. There are
+multiple other subdirectories with tests and we need to decide where to put our
+test for the evaluation of a single rule. From all these subdirectories the
+most suitable one appears to be the
+link:../../tests/API/XCCDF/unittests[tests/API/XCCDF/unittests] directory as
+all the general XCCDF tests for `oscap xccdf eval` are located there. We will
+add our new test into this subdirectory.
+
+
+==== Scenario A: There is a suitable directory for my new test
+This will happen most of the times. As stated above in our example we will place
+our test into the
+link:../../tests/API/XCCDF/unittests[tests/API/XCCDF/unittests] directory.
+
+
+==== Scenario B: There is no suitable directory for my new test
+This might happen if your test covers a part of the OpenSCAP which has no tests
+at all. In this case you would need to add a new directory with suitable name
+into the link:../../tests[tests] directory structure and create/update
+Makefiles.
+
+To have an example also for this scenario, let's suppose we want to add the
+`foo` subdirectory into the link:../../tests/API/XCCDF[tests/API/XCCDF]
+directory. We need to:
+
+. Create the `foo` subdirectory in the
+   link:../../tests/API/XCCDF[tests/API/XCCDF] directory.
+. Add a new entry for `foo` subdirectory into the `SUBDIRS` variable in
+   the link:../../tests/API/XCCDF/Makefile.am[tests/API/XCCDF/Makefile.am].
+   When adding a new entry please keep in mind that all entries are sorted in
+   the alphabetical order.
+. Create `Makefile.am` inside the `tests/API/XCCDF/foo` directory with the
+   following contents:
+[[app-listing]]
+[source,bash]
+----
+DISTCLEANFILES = *.log *.out* oscap_debug.log.*
+CLEANFILES = *.log *.out* oscap_debug.log.*
+
+TESTS_ENVIRONMENT = \
+	builddir=$(top_builddir) \
+	OSCAP_FULL_VALIDATION=1 \
+	$(top_builddir)/run
+
+# This is the entry script which should be responsible for running
+# all the tests from the 'foo' directory.
+TESTS = all.sh
+
+# This variable must contain all the files which are required to run tests.
+EXTRA_DIST = \
+	all.sh \
+	my_test_script.sh \
+	my_data.xml
+----
+
+[start=4]
+. Create an entry script, for example `all.sh`, inside the `tests/API/XCCDF/foo`
```

Comment:
We hate all.sh and are trying to move away from them. It sucks for multiple reasons, the main one is that if `all.sh` takes a long time (which some of the tests do, ~2 minutes) you can't parallelize it. If anything fails in it it's tough to navigate logs, etc etc etc...

Please provide a good example. We are aiming for a lot of small quick tests over all.sh that takes forever.